File: backend/src/test/java/com/back/ourlog/domain/timeline/service/TimelineServiceTest.java
package com.back.ourlog.domain.timeline.service;

import com.back.ourlog.domain.content.entity.Content;
import com.back.ourlog.domain.content.entity.ContentType;
import com.back.ourlog.domain.content.repository.ContentRepository;
import com.back.ourlog.domain.diary.entity.Diary;
import com.back.ourlog.domain.diary.repository.DiaryRepository;
import com.back.ourlog.domain.like.service.LikeService;
import com.back.ourlog.domain.timeline.dto.TimelineResponse;
import com.back.ourlog.domain.timeline.repository.TimelineRepository;
import com.back.ourlog.domain.user.entity.User;
import com.back.ourlog.domain.user.repository.UserRepository;
import com.back.ourlog.global.rq.Rq;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

@SpringBootTest
@Transactional
@ActiveProfiles("test")
class TimelineServiceTest {

    @Autowired
    private TimelineService timelineService;
    @Autowired private TimelineRepository timelineRepository;
    @Autowired private DiaryRepository diaryRepository;
    @Autowired private ContentRepository contentRepository;
    @Autowired private UserRepository userRepository;
    @Autowired private Rq rq;
    @Autowired private LikeService likeService;

    private User user;
    private Content content;
    private User user1;
    private User user2;

    @BeforeEach
    void setUp() {
        diaryRepository.deleteAll();
        diaryRepository.flush();
        contentRepository.deleteAll();

        user1 = userRepository.findByEmail("user1@test.com")
                .orElseGet(() -> userRepository.save(
                        User.builder()
                                .email("user1@test.com")
                                .password("pw")
                                .nickname("유저1")
                                .build()
                ));

        user2 = userRepository.findByEmail("user2@test.com")
                .orElseGet(() -> userRepository.save(
                        User.builder()
                                .email("user2@test.com")
                                .password("pw")
                                .nickname("유저2")
                                .build()
                ));

        content = contentRepository.save(new Content(
                "테스트 콘텐츠",
                ContentType.MOVIE,
                "감독이름",
                "설명",
                "https://image.com",
                LocalDateTime.now(),
                "external-id"
        ));
    }


    @Test
    @DisplayName("공개된 일기만 반환되는지 테스트")
    void testOnlyPublicDiariesReturned() {
        // given: 공개 일기와 비공개 일기 작성
        diaryRepository.save(new Diary(user1, content, "공개", "내용", 4.5f, true));
        diaryRepository.save(new Diary(user1, content, "비공개", "내용", 3.0f, false));

        // when
        List<TimelineResponse> result = timelineService.getPublicTimeline();

        // then
        assertEquals(1, result.size());
        assertEquals("공개", result.get(0).getTitle());
    }

    @DisplayName("여러 유저가 좋아요를 눌렀을 때 정확한 좋아요 수 반환")
    @Test
    void testLikeCountAccuracy() {
        // given
        Diary diary = diaryRepository.save(new Diary(user1, content, "일기 제목", "일기 내용", 4.0f, true));

        // user1이 로그인했다고 가정
        rq.setMockUser(user1);
        likeService.like(diary.getId());

        // user2이 로그인했다고 가정
        rq.setMockUser(user2);
        likeService.like(diary.getId());

        // when
        int likeCount = likeService.getLikeCount(diary.getId());

        // then
        assertEquals(2, likeCount);
    }

    @Test
    @DisplayName("댓글 수 정확성 테스트")
    void testCommentCountAccuracy() {
        // given
        Diary diary = diaryRepository.save(new Diary(user1, content, "댓글 테스트", "내용", 5.0f, true));

        // user1이 댓글 작성
        diary.addComment(user1, "첫 번째 댓글");

        // user2가 댓글 작성
        diary.addComment(user2, "두 번째 댓글");

        // 댓글을 저장하려면 diaryRepository에 저장된 diary를 다시 flush 해줘야 반영됨
        diaryRepository.save(diary);
        diaryRepository.flush();

        // when
        List<TimelineResponse> result = timelineService.getPublicTimeline();

        // then
        assertEquals(1, result.size());
        TimelineResponse timeline = result.get(0);
        assertEquals(2, timeline.getCommentCount());
    }

    @Test
    @DisplayName("로그인한 유저가 좋아요를 누른 경우 isLiked 값은 true 여야 한다")
    void testIsLikedFieldAccuracy() {
        // given
        Diary diary = diaryRepository.save(new Diary(user1, content, "일기 제목", "내용", 4.5f, true));

        rq.setMockUser(user1); // 로그인한 유저를 user1으로 설정
        likeService.like(diary.getId()); // user1이 좋아요 누름

        // when
        List<TimelineResponse> result = timelineService.getPublicTimeline();

        // then
        TimelineResponse target = result.stream()
                .filter(r -> r.getId().equals(diary.getId()))
                .findFirst()
                .orElseThrow(() -> new AssertionError("해당 일기를 찾을 수 없음"));

        assertEquals(true, target.isLiked(), "user1이 좋아요를 누른 일기에서 isLiked는 true 여야 한다");
    }

    @Test
    @DisplayName("작성자 정보가 정확하게 담겨서 반환되는지 테스트")
    void testUserSummaryAccuracy() {
        // given
        User writer = user1;
        Diary diary = diaryRepository.save(new Diary(writer, content, "타이틀", "내용", 4.0f, true));

        rq.setMockUser(writer);

        // when
        List<TimelineResponse> result = timelineService.getPublicTimeline();
        TimelineResponse response = result.get(0);

        // then
        assertEquals(writer.getId(), response.getUser().getId());
        assertEquals(writer.getNickname(), response.getUser().getNickname());
        assertEquals(writer.getProfileImageUrl(), response.getUser().getProfileImageUrl());
    }

}

File: backend/src/test/java/com/back/ourlog/domain/timeline/service/TimelineServiceTest.kt
package com.back.ourlog.domain.timeline.service

import com.back.ourlog.domain.comment.repository.CommentRepository
import com.back.ourlog.domain.content.entity.Content
import com.back.ourlog.domain.content.entity.ContentType
import com.back.ourlog.domain.diary.entity.Diary
import com.back.ourlog.domain.like.repository.LikeRepository
import com.back.ourlog.domain.timeline.repository.TimelineRepository
import com.back.ourlog.domain.user.entity.User
import com.back.ourlog.global.rq.Rq
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.junit.jupiter.api.fail
import org.mockito.BDDMockito.given
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.doReturn
import org.mockito.Mockito.spy
import org.mockito.junit.jupiter.MockitoExtension
import java.time.LocalDateTime

@ExtendWith(MockitoExtension::class)
class TimelineServiceTest {

    @Mock
    private lateinit var timelineRepository: TimelineRepository
    @Mock
    private lateinit var likeRepository: LikeRepository
    @Mock
    private lateinit var commentRepository: CommentRepository
    @Mock
    private lateinit var rq: Rq

    @InjectMocks
    private lateinit var timelineService: TimelineService

    private lateinit var user1: User
    private lateinit var content: Content
    private lateinit var diary1: Diary

    @BeforeEach
    fun setUp() {
        // 'val realUser =' 를 'user1 =' 로 수정하여 클래스 필드에 직접 할당합니다.
        user1 = User("user1@test.com", "pw", "유저1").apply { id = 1 }
        val realContent = Content(
            title = "테스트 콘텐츠", type = ContentType.MOVIE, creatorName = "감독",
            description = "설명", posterUrl = "", releasedAt = LocalDateTime.now(), externalId = "ext-1"
        )
        diary1 = Diary(user1, realContent, "공개 일기", "내용", 4.5f, true).apply {
            id = 100
            createdAt = LocalDateTime.now()
        }
    }

    @Test
    @DisplayName("공개된 일기만 반환되어야 한다")
    fun `공개된 일기만 반환되어야 한다`() {
        // given
        // 2. 이 테스트를 실행하는 데 필요한 Mock 행동만 이곳에 정의합니다.
        given(timelineRepository.findPublicDiaries()).willReturn(listOf(diary1))

        // TimelineService가 내부적으로 호출하는 다른 Mock들의 기본 동작을 설정해줍니다.
        // (이 테스트의 검증 대상은 아니지만, Null 에러 방지를 위해 필요)
        given(likeRepository.countByDiaryId(diary1.id)).willReturn(0)
        given(commentRepository.countByDiaryId(diary1.id)).willReturn(0)

        // when
        val result = timelineService.getPublicTimeline()

        // then
        assertThat(result).hasSize(1)
        assertThat(result[0].title).isEqualTo("공개 일기")
    }

    @Test
    @DisplayName("좋아요 수가 정확하게 계산되어야 한다")
    fun `좋아요 수가 정확하게 계산되어야 한다`() {
        // given
        // 1. timelineRepository가 diary1을 반환하도록 설정 (이 부분이 추가되었습니다)
        given(timelineRepository.findPublicDiaries()).willReturn(listOf(diary1))

        // 2. 이 테스트의 핵심 검증 대상인 likeRepository의 행동을 정의
        given(likeRepository.countByDiaryId(diary1.id)).willReturn(5)

        // when
        val result = timelineService.getPublicTimeline()

        // then
        assertThat(result[0].likeCount).isEqualTo(5)
    }


    @Test
    @DisplayName("댓글 수가 정확하게 계산되어야 한다")
    fun `댓글 수가 정확하게 계산되어야 한다`() {
        // given
        // 1. timelineRepository가 diary1을 반환하도록 설정 (이 부분이 추가되었습니다)
        given(timelineRepository.findPublicDiaries()).willReturn(listOf(diary1))

        // 2. 이 테스트의 핵심 검증 대상인 commentRepository의 행동을 정의
        given(commentRepository.countByDiaryId(diary1.id)).willReturn(3)

        // when
        val result = timelineService.getPublicTimeline()

        // then
        assertThat(result[0].commentCount).isEqualTo(3)
    }

    @Test
    @DisplayName("로그인한 유저가 좋아요를 누른 경우 isLiked는 true여야 한다")
    fun `로그인 유저가 좋아요 누른 경우 isLiked는 true`() {
        // given
        given(timelineRepository.findPublicDiaries()).willReturn(listOf(diary1))
        given(rq.currentUser).willReturn(user1)
        given(likeRepository.existsByUserIdAndDiaryId(user1.id, diary1.id)).willReturn(true)

        // when
        val result = timelineService.getPublicTimeline()

        // then
        assertThat(result[0].isLiked).isTrue()
    }

    @Test
    @DisplayName("작성자 정보가 정확하게 반환되어야 한다")
    fun `작성자 정보가 정확하게 반환되어야 한다`() {
        // given
        // 1. timelineRepository가 diary1을 반환하도록 설정 (이 부분이 추가되었습니다)
        given(timelineRepository.findPublicDiaries()).willReturn(listOf(diary1))

        // when
        val result = timelineService.getPublicTimeline()

        // then
        result[0].user?.let { userSummary ->
            assertThat(userSummary.id).isEqualTo(user1.id)
            assertThat(userSummary.nickname).isEqualTo(user1.nickname)
        } ?: fail("UserSummary should not be null")
    }
}